python/ray/train/_internal/storage.py

```diff
@@ -0,0 +1,236 @@
+import dataclasses
+import os
+from typing import Optional
+
+import pyarrow.fs
+
+from ray.air._internal.uri_utils import is_uri
+from ray.tune.syncer import Syncer, SyncConfig, _DefaultSyncer
+from ray.tune.result import _get_defaults_results_dir
+from ray.tune.trainable.util import TrainableUtil
+
+
+def _use_storage_context() -> bool:
+    # Whether to enable the new simple persistence mode.
+    return bool(int(os.environ.get("RAY_AIR_NEW_PERSISTENCE_MODE", "0")))
+
+
+class StorageContext:
+    """Shared context that holds all paths and storage utilities, passed along from
+    the driver to workers.
+
+    The properties of this context may not all be set at once, depending on where
+    the context lives.
+    For example, on the driver, the storage context is initialized, only knowing
+    the experiment path. On the Trainable actor, the trial_dir_name is accessible.
+
+    Example with storage_path="mock:///bucket/path":
+
+        >>> from ray.train._internal.storage import StorageContext
+        >>> import os
+        >>> os.environ["RAY_AIR_LOCAL_CACHE_DIR"] = "/tmp/ray_results"
+        >>> storage = StorageContext(
+        ...     storage_path="mock:///bucket/path",
+        ...     sync_config=SyncConfig(),
+        ...     experiment_dir_name="exp_name",
+        ... )
+        >>> storage.storage_filesystem   # Auto-resolved  # doctest: +ELLIPSIS
+        <pyarrow._fs._MockFileSystem object...
+        >>> storage.experiment_fs_path
+        'bucket/path/exp_name'
+        >>> storage.experiment_cache_path
```

Comment:
Shall we use "local_path" or "results_dir" instead of cache path as you proposed earlier? Indeed it seems confusing since it's not really caching anything.